File: ios-native/DMVCalifornia/Core/Models/Gamification.swift
import Foundation
import SwiftUI

// MARK: - User Level

struct UserLevel: Codable, Equatable {
    let level: Int
    let title: String
    let minXP: Int
    let maxXP: Int
    let icon: String
    let color: String

    var progress: Double {
        0 // Calculated externally with current XP
    }

    static let levels: [UserLevel] = [
        UserLevel(level: 1, title: "Learner", minXP: 0, maxXP: 100, icon: "leaf.fill", color: "51CF66"),
        UserLevel(level: 2, title: "Student", minXP: 100, maxXP: 250, icon: "book.fill", color: "339AF0"),
        UserLevel(level: 3, title: "Apprentice", minXP: 250, maxXP: 500, icon: "graduationcap.fill", color: "845EF7"),
        UserLevel(level: 4, title: "Scholar", minXP: 500, maxXP: 850, icon: "star.fill", color: "FCC419"),
        UserLevel(level: 5, title: "Expert", minXP: 850, maxXP: 1300, icon: "flame.fill", color: "FF922B"),
        UserLevel(level: 6, title: "Master", minXP: 1300, maxXP: 1900, icon: "bolt.fill", color: "F06595"),
        UserLevel(level: 7, title: "Champion", minXP: 1900, maxXP: 2600, icon: "trophy.fill", color: "FFD700"),
        UserLevel(level: 8, title: "Legend", minXP: 2600, maxXP: 3500, icon: "crown.fill", color: "B9F2FF"),
        UserLevel(level: 9, title: "Road Master", minXP: 3500, maxXP: 5000, icon: "car.fill", color: "FF6B6B"),
        UserLevel(level: 10, title: "DMV Pro", minXP: 5000, maxXP: Int.max, icon: "shield.fill", color: "9B59B6"),
    ]

    static func level(for xp: Int) -> UserLevel {
        levels.last { xp >= $0.minXP } ?? levels[0]
    }

    var gradientColors: [Color] {
        let baseColor = Color(hex: color)
        return [baseColor, baseColor.opacity(0.7)]
    }
}

// MARK: - XP Event

enum XPEvent: Int {
    case quizCompleted = 10
    case quizPassed = 25         // 83%+ score
    case perfectQuiz = 50        // 100% score
    case correctAnswer = 2
    case streakDay = 15
    case streakWeek = 100        // 7-day streak bonus
    case flashcardCompleted = 5
    case flashcardSetMastered = 30
    case firstQuizOfDay = 20
    case achievementUnlocked = 35

    var description: String {
        switch self {
        case .quizCompleted: return "Quiz Completed"
        case .quizPassed: return "Quiz Passed!"
        case .perfectQuiz: return "Perfect Score!"
        case .correctAnswer: return "Correct Answer"
        case .streakDay: return "Daily Streak"
        case .streakWeek: return "Week Streak Bonus"
        case .flashcardCompleted: return "Flashcard Set Done"
        case .flashcardSetMastered: return "Set Mastered"
        case .firstQuizOfDay: return "First Quiz Today"
        case .achievementUnlocked: return "Achievement!"
        }
    }
}

// MARK: - User Progress

struct UserProgress: Codable {
    var totalXP: Int = 0
    var todayXP: Int = 0
    var lastXPDate: String = ""
    var quizzesCompleted: Int = 0
    var quizzesPassed: Int = 0
    var perfectQuizzes: Int = 0
    var correctAnswers: Int = 0
    var flashcardsStudied: Int = 0
    var totalStudyTime: TimeInterval = 0
    var achievements: [String] = []

    var currentLevel: UserLevel {
        UserLevel.level(for: totalXP)
    }

    var levelProgress: Double {
        let level = currentLevel
        if level.maxXP == Int.max { return 1.0 }
        let xpInLevel = totalXP - level.minXP
        let xpNeeded = level.maxXP - level.minXP
        return min(1.0, Double(xpInLevel) / Double(xpNeeded))
    }

    var xpToNextLevel: Int {
        let level = currentLevel
        if level.maxXP == Int.max { return 0 }
        return level.maxXP - totalXP
    }
}

// MARK: - Achievement

struct Achievement: Identifiable, Codable, Equatable {
    let id: String
    let title: String
    let description: String
    let icon: String
    let tier: AchievementTier
    let category: AchievementCategory
    let requirement: Int
    var unlockedAt: Date?

    var isUnlocked: Bool {
        unlockedAt != nil
    }

    enum AchievementTier: String, Codable, CaseIterable {
        case bronze, silver, gold, platinum, diamond

        var color: Color {
            switch self {
            case .bronze: return Color(hex: "CD7F32")
            case .silver: return Color(hex: "C0C0C0")
            case .gold: return Color(hex: "FFD700")
            case .platinum: return Color(hex: "E5E4E2")
            case .diamond: return Color(hex: "B9F2FF")
            }
        }

        var xpBonus: Int {
            switch self {
            case .bronze: return 25
            case .silver: return 50
            case .gold: return 100
            case .platinum: return 200
            case .diamond: return 500
            }
        }
    }

    enum AchievementCategory: String, Codable, CaseIterable {
        case quizzes = "Quizzes"
        case streaks = "Streaks"
        case flashcards = "Flashcards"
        case mastery = "Mastery"
        case special = "Special"
    }

    // MARK: - All Achievements

    static let all: [Achievement] = [
        // Quiz Achievements
        Achievement(id: "first_quiz", title: "First Steps", description: "Complete your first quiz", icon: "flag.fill", tier: .bronze, category: .quizzes, requirement: 1),
        Achievement(id: "quiz_5", title: "Getting Started", description: "Complete 5 quizzes", icon: "doc.text.fill", tier: .bronze, category: .quizzes, requirement: 5),
        Achievement(id: "quiz_25", title: "Quiz Enthusiast", description: "Complete 25 quizzes", icon: "doc.text.fill", tier: .silver, category: .quizzes, requirement: 25),
        Achievement(id: "quiz_100", title: "Quiz Master", description: "Complete 100 quizzes", icon: "doc.text.fill", tier: .gold, category: .quizzes, requirement: 100),
        Achievement(id: "quiz_500", title: "Quiz Legend", description: "Complete 500 quizzes", icon: "doc.text.fill", tier: .diamond, category: .quizzes, requirement: 500),

        // Perfect Score Achievements
        Achievement(id: "perfect_1", title: "Perfectionist", description: "Get a perfect score", icon: "star.fill", tier: .silver, category: .mastery, requirement: 1),
        Achievement(id: "perfect_10", title: "Flawless", description: "Get 10 perfect scores", icon: "star.fill", tier: .gold, category: .mastery, requirement: 10),
        Achievement(id: "perfect_50", title: "Untouchable", description: "Get 50 perfect scores", icon: "star.fill", tier: .platinum, category: .mastery, requirement: 50),

        // Streak Achievements
        Achievement(id: "streak_3", title: "On Fire", description: "Maintain a 3-day streak", icon: "flame.fill", tier: .bronze, category: .streaks, requirement: 3),
        Achievement(id: "streak_7", title: "Week Warrior", description: "Maintain a 7-day streak", icon: "flame.fill", tier: .silver, category: .streaks, requirement: 7),
        Achievement(id: "streak_30", title: "Monthly Master", description: "Maintain a 30-day streak", icon: "flame.fill", tier: .gold, category: .streaks, requirement: 30),
        Achievement(id: "streak_100", title: "Unstoppable", description: "Maintain a 100-day streak", icon: "flame.fill", tier: .diamond, category: .streaks, requirement: 100),

        // Flashcard Achievements
        Achievement(id: "flash_set", title: "Card Collector", description: "Complete a flashcard set", icon: "rectangle.on.rectangle.fill", tier: .bronze, category: .flashcards, requirement: 1),
        Achievement(id: "flash_all", title: "Card Master", description: "Complete all flashcard sets", icon: "rectangle.on.rectangle.fill", tier: .gold, category: .flashcards, requirement: 5),

        // Special Achievements
        Achievement(id: "pass_first", title: "Ready to Drive", description: "Pass your first quiz (83%+)", icon: "car.fill", tier: .silver, category: .special, requirement: 1),
        Achievement(id: "pass_10", title: "Road Ready", description: "Pass 10 quizzes", icon: "car.fill", tier: .gold, category: .special, requirement: 10),
        Achievement(id: "level_5", title: "Rising Star", description: "Reach Level 5", icon: "arrow.up.circle.fill", tier: .silver, category: .special, requirement: 5),
        Achievement(id: "level_10", title: "DMV Pro", description: "Reach Level 10", icon: "crown.fill", tier: .diamond, category: .special, requirement: 10),
    ]

    static func get(_ id: String) -> Achievement? {
        all.first { $0.id == id }
    }
}

// MARK: - Daily Challenge

struct DailyChallenge: Identifiable, Codable {
    let id: String
    let title: String
    let description: String
    let icon: String
    let target: Int
    var current: Int
    let xpReward: Int
    let date: String

    var isCompleted: Bool {
        current >= target
    }

    var progress: Double {
        min(1.0, Double(current) / Double(target))
    }

    static func generateChallenges(for date: Date) -> [DailyChallenge] {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let dateString = dateFormatter.string(from: date)

        return [
            DailyChallenge(
                id: "daily_quiz_\(dateString)",
                title: "Quiz Time",
                description: "Complete 3 quizzes today",
                icon: "doc.text.fill",
                target: 3,
                current: 0,
                xpReward: 50,
                date: dateString
            ),
            DailyChallenge(
                id: "daily_correct_\(dateString)",
                title: "Accuracy",
                description: "Answer 20 questions correctly",
                icon: "checkmark.circle.fill",
                target: 20,
                current: 0,
                xpReward: 40,
                date: dateString
            ),
            DailyChallenge(
                id: "daily_flash_\(dateString)",
                title: "Flash Study",
                description: "Review 1 flashcard set",
                icon: "rectangle.on.rectangle.fill",
                target: 1,
                current: 0,
                xpReward: 30,
                date: dateString
            ),
        ]
    }
}

// MARK: - XP Gain Animation Data

struct XPGain: Identifiable {
    let id = UUID()
    let amount: Int
    let event: XPEvent
    let timestamp: Date = Date()
}

File: android-native/app/src/main/java/com/dmvcalifornia/app/core/model/Gamification.kt
package com.dmvcalifornia.app.core.model

import kotlinx.serialization.Serializable

// MARK: - User Level

data class UserLevel(
    val level: Int,
    val title: String,
    val minXP: Int,
    val maxXP: Int,
    val icon: String,
    val color: String
) {
    companion object {
        val levels = listOf(
            UserLevel(1, "Learner", 0, 100, "leaf", "51CF66"),
            UserLevel(2, "Student", 100, 250, "book", "339AF0"),
            UserLevel(3, "Apprentice", 250, 500, "school", "845EF7"),
            UserLevel(4, "Scholar", 500, 850, "star", "FCC419"),
            UserLevel(5, "Expert", 850, 1300, "fire", "FF922B"),
            UserLevel(6, "Master", 1300, 1900, "bolt", "F06595"),
            UserLevel(7, "Champion", 1900, 2600, "trophy", "FFD700"),
            UserLevel(8, "Legend", 2600, 3500, "crown", "B9F2FF"),
            UserLevel(9, "Road Master", 3500, 5000, "car", "FF6B6B"),
            UserLevel(10, "DMV Pro", 5000, Int.MAX_VALUE, "shield", "9B59B6"),
        )

        fun levelFor(xp: Int): UserLevel =
            levels.lastOrNull { xp >= it.minXP } ?: levels[0]
    }
}

// MARK: - XP Event

enum class XPEvent(val points: Int, val description: String) {
    QUIZ_COMPLETED(10, "Quiz Completed"),
    QUIZ_PASSED(25, "Quiz Passed!"),
    PERFECT_QUIZ(50, "Perfect Score!"),
    CORRECT_ANSWER(2, "Correct Answer"),
    STREAK_DAY(15, "Daily Streak"),
    STREAK_WEEK(100, "Week Streak Bonus"),
    FLASHCARD_COMPLETED(5, "Flashcard Set Done"),
    FLASHCARD_SET_MASTERED(30, "Set Mastered"),
    FIRST_QUIZ_OF_DAY(20, "First Quiz Today"),
    ACHIEVEMENT_UNLOCKED(35, "Achievement!");
}

// MARK: - User Progress

@Serializable
data class UserProgress(
    var totalXP: Int = 0,
    var todayXP: Int = 0,
    var lastXPDate: String = "",
    var quizzesCompleted: Int = 0,
    var quizzesPassed: Int = 0,
    var perfectQuizzes: Int = 0,
    var correctAnswers: Int = 0,
    var flashcardsStudied: Int = 0,
    var totalStudyTime: Double = 0.0,
    var achievements: MutableList<String> = mutableListOf()
) {
    val currentLevel: UserLevel get() = UserLevel.levelFor(totalXP)

    val levelProgress: Double
        get() {
            val level = currentLevel
            if (level.maxXP == Int.MAX_VALUE) return 1.0
            val xpInLevel = totalXP - level.minXP
            val xpNeeded = level.maxXP - level.minXP
            return (xpInLevel.toDouble() / xpNeeded.toDouble()).coerceAtMost(1.0)
        }

    val xpToNextLevel: Int
        get() {
            val level = currentLevel
            if (level.maxXP == Int.MAX_VALUE) return 0
            return level.maxXP - totalXP
        }
}

// MARK: - Achievement

@Serializable
data class Achievement(
    val id: String,
    val title: String,
    val description: String,
    val icon: String,
    val tier: AchievementTier,
    val category: AchievementCategory,
    val requirement: Int,
    var unlockedAt: Long? = null
) {
    val isUnlocked: Boolean get() = unlockedAt != null

    companion object {
        val all = listOf(
            // Quiz Achievements
            Achievement("first_quiz", "First Steps", "Complete your first quiz", "flag", AchievementTier.BRONZE, AchievementCategory.QUIZZES, 1),
            Achievement("quiz_5", "Getting Started", "Complete 5 quizzes", "doc_text", AchievementTier.BRONZE, AchievementCategory.QUIZZES, 5),
            Achievement("quiz_25", "Quiz Enthusiast", "Complete 25 quizzes", "doc_text", AchievementTier.SILVER, AchievementCategory.QUIZZES, 25),
            Achievement("quiz_100", "Quiz Master", "Complete 100 quizzes", "doc_text", AchievementTier.GOLD, AchievementCategory.QUIZZES, 100),
            Achievement("quiz_500", "Quiz Legend", "Complete 500 quizzes", "doc_text", AchievementTier.DIAMOND, AchievementCategory.QUIZZES, 500),
            // Perfect Score Achievements
            Achievement("perfect_1", "Perfectionist", "Get a perfect score", "star", AchievementTier.SILVER, AchievementCategory.MASTERY, 1),
            Achievement("perfect_10", "Flawless", "Get 10 perfect scores", "star", AchievementTier.GOLD, AchievementCategory.MASTERY, 10),
            Achievement("perfect_50", "Untouchable", "Get 50 perfect scores", "star", AchievementTier.PLATINUM, AchievementCategory.MASTERY, 50),
            // Streak Achievements
            Achievement("streak_3", "On Fire", "Maintain a 3-day streak", "fire", AchievementTier.BRONZE, AchievementCategory.STREAKS, 3),
            Achievement("streak_7", "Week Warrior", "Maintain a 7-day streak", "fire", AchievementTier.SILVER, AchievementCategory.STREAKS, 7),
            Achievement("streak_30", "Monthly Master", "Maintain a 30-day streak", "fire", AchievementTier.GOLD, AchievementCategory.STREAKS, 30),
            Achievement("streak_100", "Unstoppable", "Maintain a 100-day streak", "fire", AchievementTier.DIAMOND, AchievementCategory.STREAKS, 100),
            // Flashcard Achievements
            Achievement("flash_set", "Card Collector", "Complete a flashcard set", "cards", AchievementTier.BRONZE, AchievementCategory.FLASHCARDS, 1),
            Achievement("flash_all", "Card Master", "Complete all flashcard sets", "cards", AchievementTier.GOLD, AchievementCategory.FLASHCARDS, 5),
            // Special Achievements
            Achievement("pass_first", "Ready to Drive", "Pass your first quiz (83%+)", "car", AchievementTier.SILVER, AchievementCategory.SPECIAL, 1),
            Achievement("pass_10", "Road Ready", "Pass 10 quizzes", "car", AchievementTier.GOLD, AchievementCategory.SPECIAL, 10),
            Achievement("level_5", "Rising Star", "Reach Level 5", "arrow_up", AchievementTier.SILVER, AchievementCategory.SPECIAL, 5),
            Achievement("level_10", "DMV Pro", "Reach Level 10", "crown", AchievementTier.DIAMOND, AchievementCategory.SPECIAL, 10),
        )

        fun get(id: String): Achievement? = all.firstOrNull { it.id == id }
    }
}

@Serializable
enum class AchievementTier(val colorHex: String, val xpBonus: Int) {
    BRONZE("CD7F32", 25),
    SILVER("C0C0C0", 50),
    GOLD("FFD700", 100),
    PLATINUM("E5E4E2", 200),
    DIAMOND("B9F2FF", 500);
}

@Serializable
enum class AchievementCategory(val displayName: String) {
    QUIZZES("Quizzes"),
    STREAKS("Streaks"),
    FLASHCARDS("Flashcards"),
    MASTERY("Mastery"),
    SPECIAL("Special");
}

// MARK: - Daily Challenge

@Serializable
data class DailyChallenge(
    val id: String,
    val title: String,
    val description: String,
    val icon: String,
    val target: Int,
    var current: Int,
    val xpReward: Int,
    val date: String
) {
    val isCompleted: Boolean get() = current >= target
    val progress: Double get() = (current.toDouble() / target.toDouble()).coerceAtMost(1.0)

    companion object {
        fun generateChallenges(dateString: String): List<DailyChallenge> = listOf(
            DailyChallenge(
                id = "daily_quiz_$dateString",
                title = "Quiz Time",
                description = "Complete 3 quizzes today",
                icon = "doc_text",
                target = 3,
                current = 0,
                xpReward = 50,
                date = dateString
            ),
            DailyChallenge(
                id = "daily_correct_$dateString",
                title = "Accuracy",
                description = "Answer 20 questions correctly",
                icon = "check_circle",
                target = 20,
                current = 0,
                xpReward = 40,
                date = dateString
            ),
            DailyChallenge(
                id = "daily_flash_$dateString",
                title = "Flash Study",
                description = "Review 1 flashcard set",
                icon = "cards",
                target = 1,
                current = 0,
                xpReward = 30,
                date = dateString
            ),
        )
    }
}

// MARK: - XP Gain (for animation)

data class XPGain(
    val id: Long = System.nanoTime(),
    val amount: Int,
    val event: XPEvent,
    val timestamp: Long = System.currentTimeMillis()
)
